### DISCUSSION TITLE
computing mean angle

### DISCUSSION THREAD

**Discussion by axelschweiger** [pydata/xarray#6058] [Q&A] [answered]
is there a  way to create a new function to replace the  mean function in xarray for example to compute a angle mean (e.g. wind direction). If I have a python function that does this, can I tell xarray to use this in place of the built in mean? 

Thanks

**Comment by max-sixty** [comment, marked as answer]
Generally the easiest approach is to use functions rather than methods — i.e. `angle_mean = compute_angle_mean(array)`. It's possible to use duck typing in python to replace methods, but it's a difficult abstraction to manage. Xarray doesn't support swapping out functions.

**Reply by axelschweiger** [reply]
Sorry, just getting back to this. How would I use this in the context of trying to a "groupby" operation?

mWdir= hourly_data.groupby(hourly_data.time.dt.dayofyear).mean()

Do I have to do replicate everything xarray does here or can I somehow use the group object? Lookup at the groupby.apply utility which sort of sounds like it would do what I want, but can't figure out how to make this work.

**Reply by keewis** [reply]
`xarray`'s groupby objects don't have a `apply` method, but there is `.map` and `.reduce`. The difference is that `.map` gets `DataArray` / `Dataset` objects while `.reduce` works with `numpy` arrays. So if you have a function `angle_mean` that works like a `numpy` ufunc (i.e. it can be called as `angle_mean(x, axis=axis, **kwargs)`), you should be able to just pass that to `reduce`.

**Reply by axelschweiger** [reply]
Thanks much . The map function seems to work fine. See below. Any pointers on how I could expend to take the axis argument so I can use the reduce as you suggest (sorry if this is beginner stuff). 

```python
def my_mean_angle(deg):
   r=np.radians(deg) 
   x=np.sum(np.cos(r)) 
   y=np.sum(np.sin(r))
   return np.degrees(np.arctan2(y,x))

sample_data=np.linspace(0,366,366)
time=pd.date_range("1/1/2020",periods=366)
mya=xr.DataArray(sample_data,dims=['time'],coords={'time':time})
#mm=mya.groupby('time.month').mean()
mp=mya.groupby('time.month').map(my_mean_angle)
```

**Reply by keewis** [reply]
you should be able to pass it to the calls to `sum`:
```python
def my_mean_angle(deg, axis=None):
    r = np.radians(deg)
    x = np.sum(np.cos(r), axis=axis)
    y = np.sum(np.sin(r), axis=axis)
    return np.degrees(np.arctan2(y, x))

mya.groupby("time.month").reduce(my_mean_angle)
```

**Reply by axelschweiger** [reply]
Ah, simple enough, works! Interestingly, using the reduce rather than the map mechanism also allows me to us np.nansum in the my_mean_angle function which generates an error when using with map().  Strangely map() seems to be treating "nans" in sample_data as 0 even if I use np.sum....... the mysteries of xarray.   

Thanks for all your help. Much appreciated
Axel
